inc/components/class-menu.php

```diff
@@ -49,8 +61,55 @@ public function parse_wp_menu_by_location( string $menu_location ) {
 	public function parse_wp_menu( string $menu_slug ) {
 		$menu_items = wp_get_nav_menu_items( $menu_slug );
 
+		$this->children = $this->build_menu( $menu_items );
+
 		return $this;
 	}
+
+	/**
+	 * Recursive function to build a complete menu with children menu items.
+	 *
+	 * @param  array   $menu_items Menu items.
+	 * @param  integer $parent_id  Parent menu ID.
+	 * @return array All menu items.
+	 */
+	public function build_menu( $menu_items, $parent_id = 0 ) {
+
+		// Start building menu.
+		$menu = [];
+
+		// Loop through all menu items.
+		foreach ( (array) $menu_items as $key => $menu_item ) {
+
+			// Current menu's id.
+			$menu_item_id = $menu_item->ID;
+
+			// Current menu's parent id.
+			$menu_item_parent_id = absint( $menu_item->menu_item_parent );
+
+			// Is the current menu item a child of the parent item.
+			if ( $menu_item_parent_id === $parent_id ) {
+
+				// Get parsed menu item.
+				$clean_menu_item = menu_item()->parse_menu_post( $menu_item );
+
+				// Remove from loop.
+				unset( $menu_items[ $key ] );
+
+				// Recursively build children menu items.
+				$children_of_menu_item = $this->build_menu( $menu_items, $menu_item_id );
+
+				// If children menu items exist, set them as children.
+				if ( ! empty( $children_of_menu_item ) ) {
+					$clean_menu_item->set_children( $children_of_menu_item );
+				}
+
+				$menu[] = $clean_menu_item;
+			}
+		}
+
+		return (array) $menu;
+	}
```

Comment:
This all seems a bit unnecessary to me, though it's hard to tell without seeing the JS. It seems to me that this would double the net effort in building a menu: the menu items are looped-over in PHP, then inevitably JS will also have to loop over the items as well (just now in a different structure). Another approach would be to make a JS walker similar to WP's PHP walker, which would mean that the menu items are only looped-over once.

Reply:
Max and I discussed this. Since the children menu items are passed via the `children` property Irving can automatically render them in the correct nested order. We could do it on the app side, but it's just as easy to do here.